rfd/0207-vnet-ssh.md

```diff
@@ -0,0 +1,392 @@
+
+#### Connection forwarding
+
+Here is a sequence diagram for an example VNet SSH session with per-session MFA:
+
+```mermaid
+sequenceDiagram
+    participant User
+    participant Client as SSH Client
+    participant DNS_Resolver as DNS Resolver
+    participant VNet
+    participant Teleport as teleport.example.com
+    participant SSH_Server as SSH Server
+
+    User->>Client: ssh host.ssh.teleport.example.com
+    Client->>DNS_Resolver: Ask for A record for host.ssh.teleport.example.com
+    DNS_Resolver->>DNS_Resolver: Check own cache
+    DNS_Resolver->>VNet: Ask for A record for host.ssh.teleport.example.com
+    VNet->>VNet: Check own cache
+    VNet->>VNet: Apply proxy templates
+    VNet->>Teleport: Ask if SSH server "host" exists
+    Teleport->>VNet: Reply "host" exists
+    VNet->>VNet: Assign IP 100.64.0.2 to host.ssh.teleport.example.com
+    VNet->>DNS_Resolver: Return A record with IP 100.64.0.2
+    DNS_Resolver->>Client: Return A record with IP 100.64.0.2
+    Client<<->>VNet: Terminated SSH connection to 100.64.0.2
+    VNet->>Teleport: Request per-session MFA certificate
+    Teleport->>VNet: Send MFA challenge
+    VNet->>User: Prompt user to tap MFA
+    VNet->>Teleport: Send MFA response
+    Teleport->>VNet: Send MFA-verified SSH cert
+    VNet<<->>Teleport: Terminated SSH connection to "host"
+    Teleport<<->>SSH_Server: Tunnelled connection to "host"
+```
+
+### Security
+
+The point of VNet is to allow applications running on the local host to connect
+to Teleport resources transparently, without needing to configure a Teleport
+user certificate.
+It has a similar security model to a VPN, once the user securely logs in to
+Teleport on their host, all applications on the host can connect to their
+Teleport resources.
+Where Teleport improves upon the VPN security model here is with per-session
+MFA.
+When per-session MFA is enabled for SSH access in the cluster, the user will
+have to tap their MFA device every time they start a new SSH session to a
+server.
+
+#### Authentication
+
+VNet will effectively run an in-process SSH server to terminate incoming SSH
+connections and forward them to the target host.
+This SSH server will *not* require any client authentication, as mentioned above
+the point of VNet is to allow transparent access to Teleport resources by any
+application running on the local host where the user has logged in.
+
+The VNet in-process SSH server will set
+[`ssh.ServerConfig.NoClientAuth`](https://pkg.go.dev/golang.org/x/crypto/ssh#ServerConfig)
+to `true` to disable client authentication.
+However, it will also set `ssh.ServerConfig.NoClientAuthCallback`.
+In this callback is where it will attempt to dial the target server with the
+requested username, and complete the per-session MFA ceremony if necessary.
+By doing this in the callback we can return an authentication error to the user
+if they fail to authenticate to the target server.
```

Comment:
This is more of an implementation detail that doesn't necessarily need to be mentioned in the RFD, but this actually supports dropping into `keyboard-interactive` if the remote server is configured for `publickey,keyboard-interactive` in a much better way than our current forwarding (some customers have asked about better support for PAM authentication modules, and `x/crypto/ssh` doesn't quite have APIs flexible enough for that if the connecting client is trying to use `publickey` first).